File: python/mlx-movie-director/app/controlnet.py
"""MLX ControlNet model for Z-Image Turbo (Union 2.0).

Source: https://civitai.com/models/2192289/zimageturbo-controlnet-6g-vram-can-run-it?modelVersionId=2509261

Architecture (295 keys in safetensors):
  control_all_x_embedder["2-1"]  [3840, 132]  — embeds concat(noise[64], ctrl[64], type[4])
  control_noise_refiner[0,1]                  — 2 refinement transformer blocks
  control_layers[0..14]                       — 15 blocks with before_proj + after_proj

The 15 control_layer outputs inject into the main ZImageTransformerMLX layers at stride-2:
  control_layers[i] residual → main transformer layer[2*i]

Module attribute names mirror safetensors key paths exactly to enable direct weight loading.
"""

import os
import sys

import mlx.core as mx
import mlx.nn as nn


_DIM = 3840
_N_HEADS = 30
_HEAD_DIM = _DIM // _N_HEADS   # 128
_HIDDEN = int(_DIM / 3 * 8)    # 10240
_EMBED_DIM = 64                 # patch_size^2 * channels = 2^2 * 16
_UNION_DIM = 4
_CONTROL_IN = _EMBED_DIM + _EMBED_DIM + _UNION_DIM  # 132
_T_EMB_DIM = 256               # min(dim, 256) from main transformer
_N_CONTROL_LAYERS = 15
_N_REFINERS = 2


# ---------------------------------------------------------------------------
# Sub-modules (attribute names match safetensors key structure)
# ---------------------------------------------------------------------------

class _ControlNetAttention(nn.Module):
    def __init__(self):
        super().__init__()
        self.to_q = nn.Linear(_DIM, _DIM, bias=False)
        self.to_k = nn.Linear(_DIM, _DIM, bias=False)
        self.to_v = nn.Linear(_DIM, _DIM, bias=False)
        # List → parameter name is attention.to_out.0.* matching safetensors keys
        self.to_out = [nn.Linear(_DIM, _DIM, bias=False)]
        self.norm_q = nn.RMSNorm(_HEAD_DIM, eps=1e-5)
        self.norm_k = nn.RMSNorm(_HEAD_DIM, eps=1e-5)
        self.scale = _HEAD_DIM ** -0.5

    def __call__(self, x, cos, sin):
        B, L, D = x.shape
        q = self.to_q(x).reshape(B, L, _N_HEADS, _HEAD_DIM)
        k = self.to_k(x).reshape(B, L, _N_HEADS, _HEAD_DIM)
        v = self.to_v(x).reshape(B, L, _N_HEADS, _HEAD_DIM)
        q = self.norm_q(q)
        k = self.norm_k(k)
        if cos is not None and sin is not None:
            q1, q2 = q[..., 0::2], q[..., 1::2]
            q = mx.stack([q1 * cos - q2 * sin, q1 * sin + q2 * cos], axis=-1).reshape(B, L, _N_HEADS, _HEAD_DIM)
            k1, k2 = k[..., 0::2], k[..., 1::2]
            k = mx.stack([k1 * cos - k2 * sin, k1 * sin + k2 * cos], axis=-1).reshape(B, L, _N_HEADS, _HEAD_DIM)
        q = q.transpose(0, 2, 1, 3)
        k = k.transpose(0, 2, 1, 3)
        v = v.transpose(0, 2, 1, 3)
        out = mx.fast.scaled_dot_product_attention(q, k, v, scale=self.scale, mask=None)
        out = out.transpose(0, 2, 1, 3).reshape(B, L, D)
        return self.to_out[0](out)


class _ControlNetFeedForward(nn.Module):
    def __init__(self):
        super().__init__()
        self.w1 = nn.Linear(_DIM, _HIDDEN, bias=False)
        self.w2 = nn.Linear(_HIDDEN, _DIM, bias=False)
        self.w3 = nn.Linear(_DIM, _HIDDEN, bias=False)

    def __call__(self, x):
        return self.w2(nn.silu(self.w1(x)) * self.w3(x))


def _block_forward(block, x, temb, cos, sin):
    """Shared attention+FFN+modulation forward used by all control blocks."""
    chunks = block.adaLN_modulation[0](temb)
    scale_msa, gate_msa, scale_mlp, gate_mlp = mx.split(chunks, 4, axis=-1)
    scale_msa = (1.0 + scale_msa)[:, None, :]
    scale_mlp = (1.0 + scale_mlp)[:, None, :]
    gate_msa = mx.tanh(gate_msa)[:, None, :]
    gate_mlp = mx.tanh(gate_mlp)[:, None, :]
    attn_out = block.attention(block.attention_norm1(x) * scale_msa, cos, sin)
    x = x + gate_msa * block.attention_norm2(attn_out)
    ffn_out = block.feed_forward(block.ffn_norm1(x) * scale_mlp)
    x = x + gate_mlp * block.ffn_norm2(ffn_out)
    return x


class _ControlNetBlockFirst(nn.Module):
    """Index-0 block — has before_proj AND after_proj (matches safetensors key structure)."""
    def __init__(self):
        super().__init__()
        self.adaLN_modulation = [nn.Linear(_T_EMB_DIM, 4 * _DIM, bias=True)]
        self.attention = _ControlNetAttention()
        self.feed_forward = _ControlNetFeedForward()
        self.attention_norm1 = nn.RMSNorm(_DIM, eps=1e-5)
        self.attention_norm2 = nn.RMSNorm(_DIM, eps=1e-5)
        self.ffn_norm1 = nn.RMSNorm(_DIM, eps=1e-5)
        self.ffn_norm2 = nn.RMSNorm(_DIM, eps=1e-5)
        self.before_proj = nn.Linear(_DIM, _DIM, bias=True)
        self.after_proj = nn.Linear(_DIM, _DIM, bias=True)

    def __call__(self, x, temb, cos, sin):
        """Return (hidden_state, residual)."""
        h = self.before_proj(x)
        h = _block_forward(self, h, temb, cos, sin)
        return h, self.after_proj(h)


class _ControlNetBlockRest(nn.Module):
    """Index > 0 block — has only after_proj, NO before_proj (matches safetensors key structure)."""
    def __init__(self):
        super().__init__()
        self.adaLN_modulation = [nn.Linear(_T_EMB_DIM, 4 * _DIM, bias=True)]
        self.attention = _ControlNetAttention()
        self.feed_forward = _ControlNetFeedForward()
        self.attention_norm1 = nn.RMSNorm(_DIM, eps=1e-5)
        self.attention_norm2 = nn.RMSNorm(_DIM, eps=1e-5)
        self.ffn_norm1 = nn.RMSNorm(_DIM, eps=1e-5)
        self.ffn_norm2 = nn.RMSNorm(_DIM, eps=1e-5)
        self.after_proj = nn.Linear(_DIM, _DIM, bias=True)

    def __call__(self, x, temb, cos, sin):
        """Return (hidden_state, residual)."""
        h = _block_forward(self, x, temb, cos, sin)
        return h, self.after_proj(h)


# ---------------------------------------------------------------------------
# Full ControlNet model
# ---------------------------------------------------------------------------

class ZImageControlnet(nn.Module):
    """Z-Image Union ControlNet 2.0 native MLX implementation."""

    def __init__(self):
        super().__init__()
        # Dict key "2-1" → parameter path control_all_x_embedder.2-1.* (matches safetensors)
        # Dict key "2-1" → path control_all_x_embedder.2-1.* (matches safetensors)
        self.control_all_x_embedder = {"2-1": nn.Linear(_CONTROL_IN, _DIM, bias=True)}
        # Index-0 blocks have before_proj; subsequent blocks do not (matches safetensors)
        self.control_noise_refiner = [
            _ControlNetBlockFirst(),
            *[_ControlNetBlockRest() for _ in range(_N_REFINERS - 1)],
        ]
        self.control_layers = [
            _ControlNetBlockFirst(),
            *[_ControlNetBlockRest() for _ in range(_N_CONTROL_LAYERS - 1)],
        ]

    def __call__(
        self,
        x_raw: mx.array,       # [1, N_padded, 64]  patchified noise latent
        ctrl_raw: mx.array,    # [1, N_padded, 64]  patchified control latent
        union_type: int,        # 0=pose 1=depth 2=canny 3=hed 4=scribble
        temb: mx.array,         # [1, 256]  time embedding from main transformer
        cos: mx.array,          # [1, N_padded, 1, 64]  rope cos (image token slice)
        sin: mx.array,          # [1, N_padded, 1, 64]  rope sin (image token slice)
        strength: float = 1.0,
    ) -> list:
        """Return list of 15 residual tensors [1, N_padded, 3840] scaled by strength."""
        B, N, _ = x_raw.shape

        # One-hot union type vector, broadcast to [B, N, 4]
        idx = union_type % _UNION_DIM
        type_onehot = mx.array(
            [1.0 if i == idx else 0.0 for i in range(_UNION_DIM)],
            dtype=x_raw.dtype,
        )
        type_vec = mx.broadcast_to(type_onehot[None, None, :], (B, N, _UNION_DIM))

        # Embed: [B, N, 64+64+4] → [B, N, 3840]
        combined = mx.concatenate([x_raw, ctrl_raw, type_vec], axis=-1)
        x = self.control_all_x_embedder["2-1"](combined)

        # Noise refiners: advance hidden state, discard residuals
        for block in self.control_noise_refiner:
            x, _ = block(x, temb, cos, sin)

        # Control layers: collect residuals for injection into main transformer
        residuals = []
        for block in self.control_layers:
            x, residual = block(x, temb, cos, sin)
            residuals.append(residual * strength)

        return residuals   # 15 × [1, N_padded, 3840]


# ---------------------------------------------------------------------------
# Weight loader
# ---------------------------------------------------------------------------

def load_controlnet(model_dir: str) -> ZImageControlnet:
    """Load ZImageControlnet weights from model_dir/model.safetensors."""
    model_path = os.path.join(model_dir, "model.safetensors")
    if not os.path.exists(model_path):
        raise FileNotFoundError(f"ControlNet weights not found: {model_path}")

    print(f"[ControlNet] Loading {os.path.getsize(model_path) // 1_000_000} MB from {model_path}...", end=" ", flush=True)
    weights = mx.load(model_path)
    print(f"({len(weights)} tensors)", flush=True)

    model = ZImageControlnet()
    model.load_weights(list(weights.items()))
    mx.eval(model.parameters())
    print("[ControlNet] Ready.")
    return model


# ---------------------------------------------------------------------------
# Patchification helper (mirrors ZImagePipeline.generate() step_fn logic)
# ---------------------------------------------------------------------------

def patchify_latent(latent: mx.array) -> mx.array:
    """Patchify latent [1, C, H, W] → patches [1, H_tok*W_tok, C*4].

    Matches the reshape in pipeline.py step_fn so that N_patches is the same
    for noise and control latents (both must have the same H, W before calling).
    """
    B, C, H, W = latent.shape
    # Delegate to single-batch version for consistency with step_fn
    x = latent[0]                       # [C, H, W]
    H_tok, W_tok = H // 2, W // 2
    x = x.reshape(C, 1, 1, H_tok, 2, W_tok, 2)
    x = x.transpose(1, 2, 3, 5, 4, 6, 0)   # [1, 1, H_tok, W_tok, 2, 2, C]
    x = x.reshape(1, H_tok * W_tok, C * 4)  # [1, N, 64]
    return x

File: swift/z-image-director/Sources/ZImageDirector/ControlNet.swift
//
//  ControlNet.swift
//  ZImageDirector
//
//  Z-Image Union ControlNet 2.1 (native MLX port of app/controlnet.py).
//  Turbo-specific: pairs with the Z-Image Turbo transformer (moody-pro-mix etc).
//
//  Architecture (185 quantized keys):
//    control_all_x_embedder["2-1"]   [3840, 132]  — embeds 33-ch control image patches
//    control_noise_refiner[0,1]                   — 2 refinement blocks (block 0 has before_proj)
//    control_layers[0..2]                         — 3 blocks (block 0 has before_proj)
//
//  The control blocks run interleaved with the main transformer's forward pass,
//  receiving the main model's hidden state at each injection point. Residuals
//  are added back into the main image tokens × controlnet_strength.
//
//  Input: 33-channel patchified image [control_latent(16), mask(1), inpaint_latent(16)]
//    For non-inpaint: mask=zeros, inpaint_latent=VAE(gray_image).
//

import Foundation
import MLX
import MLXNN
import MLXFast

// Architecture constants (mirror app/controlnet.py)
private let CN_DIM = 3840
private let CN_N_HEADS = 30
private let CN_HEAD_DIM = CN_DIM / CN_N_HEADS   // 128
private let CN_HIDDEN = Int(CN_DIM / 3 * 8)     // 10240
private let CN_T_EMB_DIM = 256                  // min(dim, 256) from main transformer
private let CN_CONTROL_IN_DIM = 132             // 33 ch × 4 (patchification)
private let CN_N_CONTROL_LAYERS = 3
private let CN_N_REFINERS = 2

/// Flux latent format (ZImage inherits → Flux VAE).
/// process_in(latent) = (latent - shift) * scale. Matches _FLUX_* in controlnet.py.
public enum FluxLatent {
    public static let shiftFactor: Float = 0.1159
    public static let scaleFactor: Float = 0.3611
}

// MARK: - ControlNet sub-modules

/// Attention for control blocks. to_q/k/v/out + RMSNorm on q,k. No bias.
/// Mirrors _ControlNetAttention; reuses MLXFast.scaledDotProductAttention.
final class ControlNetAttention: Module {
    let toq: QuantizedLinear
    let tok: QuantizedLinear
    let tov: QuantizedLinear
    let too: QuantizedLinear
    let normQ: ZRMSNorm
    let normK: ZRMSNorm
    let scale: Float

    init(toq: QuantizedLinear, tok: QuantizedLinear, tov: QuantizedLinear,
         too: QuantizedLinear, normQ: ZRMSNorm, normK: ZRMSNorm) {
        self.toq = toq; self.tok = tok; self.tov = tov; self.too = too
        self.normQ = normQ; self.normK = normK
        self.scale = pow(Float(CN_HEAD_DIM), -0.5)
        super.init()
    }

    func callAsFunction(_ x: MLXArray, cos: MLXArray?, sin: MLXArray?) -> MLXArray {
        let (B, L, _) = (x.dim(0), x.dim(1), x.dim(2))
        var q = toq(x).reshaped([B, L, CN_N_HEADS, CN_HEAD_DIM])
        var k = tok(x).reshaped([B, L, CN_N_HEADS, CN_HEAD_DIM])
        var v = tov(x).reshaped([B, L, CN_N_HEADS, CN_HEAD_DIM])
        q = normQ(q)
        k = normK(k)
        if let cos = cos, let sin = sin {
            // RoPE: rotate even/odd index pairs. cos/sin are (B, L, 1, headDim/2).
            let half = CN_HEAD_DIM / 2
            let qr = q.reshaped([B, L, CN_N_HEADS, half, 2])
            let q1 = qr[0..., 0..., 0..., 0..., 0]; let q2 = qr[0..., 0..., 0..., 0..., 1]
            q = MLX.stacked([q1 * cos - q2 * sin, q1 * sin + q2 * cos], axis: -1)
                .reshaped([B, L, CN_N_HEADS, CN_HEAD_DIM])
            let kr = k.reshaped([B, L, CN_N_HEADS, half, 2])
            let k1 = kr[0..., 0..., 0..., 0..., 0]; let k2 = kr[0..., 0..., 0..., 0..., 1]
            k = MLX.stacked([k1 * cos - k2 * sin, k1 * sin + k2 * cos], axis: -1)
                .reshaped([B, L, CN_N_HEADS, CN_HEAD_DIM])
        }
        q = q.transposed(0, 2, 1, 3)  // (B, nheads, L, headDim)
        k = k.transposed(0, 2, 1, 3)
        v = v.transposed(0, 2, 1, 3)
        let out = MLXFast.scaledDotProductAttention(
            queries: q, keys: k, values: v, scale: scale, mask: nil)
        return too(out.reshaped([B, L, CN_DIM]))
    }
}

/// Feed-forward (SwiGLU): w2(silu(w1(x)) * w3(x)). Mirrors _ControlNetFeedForward.
final class ControlNetFeedForward: Module {
    let w1: QuantizedLinear
    let w2: QuantizedLinear
    let w3: QuantizedLinear

    init(w1: QuantizedLinear, w2: QuantizedLinear, w3: QuantizedLinear) {
        self.w1 = w1; self.w2 = w2; self.w3 = w3
        super.init()
    }

    func callAsFunction(_ x: MLXArray) -> MLXArray {
        w2(MLXNN.silu(w1(x)) * w3(x))
    }
}

/// Shared attention+FFN+modulation forward for control blocks.
/// Mirrors _block_forward in controlnet.py.
final class ControlNetCore {
    let adaLN: QuantizedLinear           // adaLN_modulation.0: 256 → 4*dim
    let attention: ControlNetAttention
    let feedForward: ControlNetFeedForward
    let attentionNorm1: ZRMSNorm
    let attentionNorm2: ZRMSNorm
    let ffnNorm1: ZRMSNorm
    let ffnNorm2: ZRMSNorm

    init(adaLN: QuantizedLinear, attention: ControlNetAttention,
         feedForward: ControlNetFeedForward,
         attentionNorm1: ZRMSNorm, attentionNorm2: ZRMSNorm,
         ffnNorm1: ZRMSNorm, ffnNorm2: ZRMSNorm) {
        self.adaLN = adaLN; self.attention = attention; self.feedForward = feedForward
        self.attentionNorm1 = attentionNorm1; self.attentionNorm2 = attentionNorm2
        self.ffnNorm1 = ffnNorm1; self.ffnNorm2 = ffnNorm2
    }

    /// Run the attention+FFN+modulation core (no before/after projection).
    func core(_ h: MLXArray, temb: MLXArray, cos: MLXArray?, sin: MLXArray?) -> MLXArray {
        let chunks = adaLN(temb)
        let parts = chunks.split(parts: 4, axis: -1)
        var scaleMsa = parts[0]; var gateMsa = parts[1]
        var scaleMlp = parts[2]; var gateMlp = parts[3]
        scaleMsa = (1.0 + scaleMsa).expandedDimensions(axis: -2)
        scaleMlp = (1.0 + scaleMlp).expandedDimensions(axis: -2)
        gateMsa = MLX.tanh(gateMsa).expandedDimensions(axis: -2)
        gateMlp = MLX.tanh(gateMlp).expandedDimensions(axis: -2)
        let attnOut = attention(attentionNorm1(h) * scaleMsa, cos: cos, sin: sin)
        var x = h + gateMsa * attentionNorm2(attnOut)
        let ffnOut = feedForward(ffnNorm1(x) * scaleMlp)
        x = x + gateMlp * ffnNorm2(ffnOut)
        return x
    }
}

/// Index-0 block — has before_proj AND after_proj.
final class ControlNetBlockFirst {
    let core: ControlNetCore
    let beforeProj: QuantizedLinear
    let afterProj: QuantizedLinear

    init(core: ControlNetCore, beforeProj: QuantizedLinear, afterProj: QuantizedLinear) {
        self.core = core; self.beforeProj = beforeProj; self.afterProj = afterProj
    }

    /// Returns (residual, updatedContext).
    /// At block 0: project control context and ADD main model's hidden state.
    func callAsFunction(
        _ controlContext: MLXArray, mainHidden: MLXArray,
        temb: MLXArray, cos: MLXArray?, sin: MLXArray?
    ) -> (MLXArray, MLXArray) {
        let h = beforeProj(controlContext) + mainHidden
        let out = core.core(h, temb: temb, cos: cos, sin: sin)
        return (afterProj(out), out)
    }
}

/// Index > 0 block — only after_proj (main hidden NOT added).
final class ControlNetBlockRest {
    let core: ControlNetCore
    let afterProj: QuantizedLinear

    init(core: ControlNetCore, afterProj: QuantizedLinear) {
        self.core = core; self.afterProj = afterProj
    }

    func callAsFunction(
        _ controlContext: MLXArray, mainHidden: MLXArray,
        temb: MLXArray, cos: MLXArray?, sin: MLXArray?
    ) -> (MLXArray, MLXArray) {
        let out = core.core(controlContext, temb: temb, cos: cos, sin: sin)
        return (afterProj(out), out)
    }
}

// MARK: - Full ControlNet

/// Z-Image Union ControlNet 2.1. Supports interleaved execution where control
/// blocks receive the main model's hidden state at each injection point.
public final class ZImageControlNet: Module {

    let controlXEmbedder: Linear                // control_all_x_embedder["2-1"]: 132 → 3880 (bf16, unquantized)
    let controlNoiseRefiner: [Any]                 // [ControlNetBlockFirst, ControlNetBlockRest]
    let controlLayers: [Any]                       // [First, Rest, Rest]

    /// Load from a weights dict (keys prefixed as in safetensors).
    public init(weights: [String: MLXArray], groupSize: Int = 32, bits: Int = 4) {
        // Quantized linear: weight + scales + biases (+ optional bias).
        func ql(_ prefix: String) -> QuantizedLinear {
            QuantizedLinear(
                weight: weights[prefix + ".weight"]!,
                bias: weights[prefix + ".bias"],
                scales: weights[prefix + ".scales"]!,
                biases: weights[prefix + ".biases"]!,
                groupSize: groupSize, bits: bits
            )
        }
        // Unquantized linear: weight + bias (no scales/biases). The embedder
        // ships as bf16; we wrap it in a plain Linear (callAsFunction identical).
        func linear(_ prefix: String) -> Linear {
            Linear(weight: weights[prefix + ".weight"]!, bias: weights[prefix + ".bias"])
        }
        func rms(_ prefix: String) -> ZRMSNorm {
            ZRMSNorm(weight: weights[prefix + ".weight"]!, eps: 1e-5)
        }
        func core(_ prefix: String) -> ControlNetCore {
            ControlNetCore(
                adaLN: ql("\(prefix).adaLN_modulation.0"),
                attention: ControlNetAttention(
                    toq: ql("\(prefix).attention.to_q"),
                    tok: ql("\(prefix).attention.to_k"),
                    tov: ql("\(prefix).attention.to_v"),
                    too: ql("\(prefix).attention.to_out.0"),
                    normQ: rms("\(prefix).attention.norm_q"),
                    normK: rms("\(prefix).attention.norm_k")
                ),
                feedForward: ControlNetFeedForward(
                    w1: ql("\(prefix).feed_forward.w1"),
                    w2: ql("\(prefix).feed_forward.w2"),
                    w3: ql("\(prefix).feed_forward.w3")
                ),
                attentionNorm1: rms("\(prefix).attention_norm1"),
                attentionNorm2: rms("\(prefix).attention_norm2"),
                ffnNorm1: rms("\(prefix).ffn_norm1"),
                ffnNorm2: rms("\(prefix).ffn_norm2")
            )
        }
        func blockFirst(_ prefix: String) -> ControlNetBlockFirst {
            ControlNetBlockFirst(
                core: core(prefix),
                beforeProj: ql("\(prefix).before_proj"),
                afterProj: ql("\(prefix).after_proj")
            )
        }
        func blockRest(_ prefix: String) -> ControlNetBlockRest {
            ControlNetBlockRest(core: core(prefix), afterProj: ql("\(prefix).after_proj"))
        }

        self.controlXEmbedder = linear("control_all_x_embedder.2-1")
        self.controlNoiseRefiner = [
            blockFirst("control_noise_refiner.0"),
            blockRest("control_noise_refiner.1"),
        ]
        self.controlLayers = [
            blockFirst("control_layers.0"),
            blockRest("control_layers.1"),
            blockRest("control_layers.2"),
        ]
        super.init()
    }

    /// Embed the 33-channel control image into the ControlNet hidden space.
    /// Input [1, 33, H, W] → [1, N_patches, 3840].
    public func embedControl(_ controlImage33ch: MLXArray) -> MLXArray {
        let (B, C, H, W) = (controlImage33ch.dim(0), controlImageShape1(controlImage33ch),
                            controlImage33ch.dim(2), controlImage33ch.dim(3))
        _ = (B, C)
        let patch = 2
        let hTok = H / patch, wTok = W / patch
        // [B, C, H, W] → [B, C, hTok, 2, wTok, 2] → [B, hTok, wTok, 2, 2, C] → [B, N, 132]
        var x = controlImage33ch.reshaped([B, C, hTok, patch, wTok, patch])
        x = x.transposed(0, 2, 4, 3, 5, 1)
        x = x.reshaped([B, hTok * wTok, C * patch * patch])
        return controlXEmbedder(x)
    }

    private func controlImageShape1(_ a: MLXArray) -> Int { a.dim(1) }

    /// Run one noise refiner control block.
    /// Returns (residual, updatedContext).
    public func forwardNoiseRefiner(
        layerId: Int, controlContext: MLXArray, mainHidden: MLXArray,
        temb: MLXArray, cos: MLXArray?, sin: MLXArray?
    ) -> (MLXArray, MLXArray) {
        if layerId == 0 {
            return (controlNoiseRefiner[0] as! ControlNetBlockFirst)(
                controlContext, mainHidden: mainHidden, temb: temb, cos: cos, sin: sin)
        } else {
            return (controlNoiseRefiner[layerId] as! ControlNetBlockRest)(
                controlContext, mainHidden: mainHidden, temb: temb, cos: cos, sin: sin)
        }
    }

    /// Run one control layer block (stride-2 injection into main layers).
    public func forwardControlLayer(
        layerId: Int, controlContext: MLXArray, mainHidden: MLXArray,
        temb: MLXArray, cos: MLXArray?, sin: MLXArray?
    ) -> (MLXArray, MLXArray) {
        if layerId == 0 {
            return (controlLayers[0] as! ControlNetBlockFirst)(
                controlContext, mainHidden: mainHidden, temb: temb, cos: cos, sin: sin)
        } else {
            return (controlLayers[layerId] as! ControlNetBlockRest)(
                controlContext, mainHidden: mainHidden, temb: temb, cos: cos, sin: sin)
        }
    }

    public var nControlLayers: Int { controlLayers.count }
}

// MARK: - Control input builder

public enum ControlNetInput {
    /// Build the 33-channel control input: [control_latent(16), mask(1), inpaint_latent(16)].
    /// All inputs are Flux-normalized latents [1, C, H, W].
    ///
    /// - Parameters:
    ///   - ctrlLatent: [1, 16, H, W] VAE-encoded control image (already Flux-normalized).
    ///   - inpaintLatent: optional [1, 16, H, W] source-image latent for appearance
    ///     anchor (Flux-normalized). nil → zeros (no inpaint anchor).
    ///   - mask: optional [1, 1, H, W] spatial mask. 0=preserve source, 1=allow
    ///     generation. nil → uniform zeros (no inpaint guidance).
    public static func build33ch(
        ctrlLatent: MLXArray, inpaintLatent: MLXArray? = nil, mask: MLXArray? = nil
    ) -> MLXArray {
        let H = ctrlLatent.dim(2)
        let W = ctrlLatent.dim(3)
        let dt = ctrlLatent.dtype
        let inpaint = inpaintLatent ?? MLX.zeros([1, 16, H, W]).asType(dt)
        let m = mask ?? MLX.zeros([1, 1, H, W]).asType(dt)
        return MLX.concatenated([ctrlLatent, m.asType(dt), inpaint.asType(dt)], axis: 1)
    }

    /// Convenience: build from a control PIL image + pipeline encoder.
    /// Encodes the control image and a neutral gray inpaint anchor.
    public static func build33chFromImage(
        controlPixels: MLXArray, pipeline: T2IPipeline,
        inpaintPixels: MLXArray? = nil, mask: MLXArray? = nil
    ) -> MLXArray {
        let ctrlNorm = ImageLoad.normalizeForVAE(controlPixels)
        let ctrlLatent = pipeline.encodeImage(ctrlNorm)
        let inpaintLatent: MLXArray?
        if let inp = inpaintPixels {
            inpaintLatent = pipeline.encodeImage(ImageLoad.normalizeForVAE(inp))
        } else {
            // Neutral gray anchor: encode mid-gray so the inpaint channel is neutral.
            let (W, H) = (controlPixels.dim(3), controlPixels.dim(2))
            let gray = MLX.full([1, 3, H, W], values: 0.5).asType(controlPixels.dtype)
            inpaintLatent = pipeline.encodeImage(ImageLoad.normalizeForVAE(gray))
        }
        return build33ch(ctrlLatent: ctrlLatent, inpaintLatent: inpaintLatent, mask: mask)
    }
}
